src/nvim/ui.c

```diff
@@ -520,7 +521,7 @@ void ui_flush(void)
     api_free_array(style);
     pending_mode_info_update = false;
   }
-  if (pending_mode_update) {
+  if (pending_mode_update && !starting) {
```

Comment:
why are we only guarding `pending_mode_update` and not the other potential updates? 

Maybe we need to delay `ui_flush()` entirely?

Reply:
Based on this comment from @erw7 https://github.com/neovim/neovim/issues/18470#issuecomment-1126636767, my understanding is that this issue is caused by mode transition occured incorrectly during the ui_flush (due to node changing the state of the tty on exit, according to that comment). So we should only need to guard against `pending_mode_update`, and I think delaying `ui_flush` would not solve this issue because the state change to the tty caused by node would still remain.

Reply:
Delaying `ui_flush()` also solves this problem. The reason I haven't done so in the https://github.com/neovim/neovim/issues/18470#issuecomment-1153117726 patch is that I'm not sure that delaying the entire `ui_flush()` would cause any other problems. This change is almost a workaround. While it may be better to create a new event and send it to the UI with `VimEnter`, the reason I suggested this simplified method is due to the tradeoff between the importance of the problem being solved and the effort involved.

Reply:
> tradeoff between the importance of the problem being solved and the effort involved.

:+1: also "remote/coprocess TUI" https://github.com/neovim/neovim/pull/18375 changes the architecture entirely, so best to wait on that.